hir/src/asm.rs

```diff
@@ -0,0 +1,2447 @@
+    /// remain on the operand stack; if `Type::Felt` is given, then a single value should be on the operand stack; if
+    /// `Type::Array[Type::Felt; 2]` is given, then two values should be on the operand stack, and so on.
+    ///
+    /// NOTE: Not all types are permitted as inline assembly results. The type must be "loadable", i.e. no larger than a word.
+    ///
+    /// Any attempt to modify the operand stack beyond what is made visible via arguments, or introduced within the
+    /// inline assembly block, will cause an assertion to fail.
+    pub fn new(mut builder: B, args: &[Value], ty: Type, span: SourceSpan) -> Self {
+        assert!(
+            ty.is_loadable(),
+            "invalid inline assembly block type: type must be loadable, but got {}",
+            &ty
+        );
+        // Construct the initial operand stack with the given arguments
+        let mut stack = OperandStack::<Type>::default();
+        {
+            let dfg = builder.data_flow_graph();
+            for arg in args.iter().rev().copied() {
+                stack.push(dfg.value_type(arg).clone());
+            }
+        }
+
+        // Construct an empty inline assembly block with the given arguments
+        let mut asm = InlineAsm::new();
+        {
+            let dfg = builder.data_flow_graph_mut();
+            let mut vlist = ValueList::default();
+            vlist.extend(args.iter().copied(), &mut dfg.value_lists);
+            asm.args = vlist;
+        }
+
+        let current_block = asm.body;
+        Self {
+            builder,
+            span,
+            asm,
+            ty,
+            current_block,
+            stack,
+        }
+    }
+
+    #[inline]
+    pub fn create_block(&mut self) -> MasmBlockId {
+        self.asm.create_block()
+    }
+
+    #[inline(always)]
+    pub fn switch_to_block(&mut self, block: MasmBlockId) {
+        self.current_block = block;
+    }
+
+    pub fn ins<'a, 'b: 'a>(&'b mut self) -> MasmOpBuilder<'a> {
+        MasmOpBuilder {
+            asm: &mut self.asm,
+            stack: &mut self.stack,
+            ip: self.current_block,
+        }
+    }
+
+    pub fn build(self) -> (Inst, &'f mut DataFlowGraph) {
+        let ty = self.ty;
+        match &ty {
+            Type::Unit => assert!(self.stack.is_empty(), "invalid inline assembly: expected operand stack to be empty upon exit, found: {:?}", self.stack.display()),
+            ty => {
+                let len = ty.size_in_felts();
+                assert_eq!(len, self.stack.len(), "invalid inline assembly: expected operand stack to have {} elements upon exit, found: {:?}", len, self.stack.display());
+            }
+        }
+
+        let span = self.span;
+        let data = Instruction::InlineAsm(self.asm);
+        self.builder.build(data, ty, span)
+    }
+}
+
+/// Used to construct a single MASM opcode
+pub struct MasmOpBuilder<'a> {
+    asm: &'a mut InlineAsm,
+    stack: &'a mut OperandStack<Type>,
```

Comment:
I'm not sure of the purpose of the `stack` field. In particular, I'm not sure why the stack contains `Type` objects. Please add a comment.